index.js

```diff
@@ -1,3 +1,45 @@
-require.addon = require('require-addon')
+const binding = require('./binding')
+const { isNode } = require('which-runtime')
 
-module.exports = require.addon('.', __filename)
+module.exports = binding
+
+module.exports.sodium_malloc = size => {
+  const buf = Buffer.from(binding._sodium_malloc(size))
+  buf.secure = true
+  return buf
+}
+
+module.exports.crypto_generichash = function (output, input, key = undefined) {
+  return binding._crypto_generichash(output, input, key)
+}
+
+module.exports.crypto_generichash_init = function (state, key, outputLength) {
+  if (state.byteLength !== binding.crypto_generichash_STATEBYTES) throw new Error("state must be 'crypto_generichash_STATEBYTES' bytes")
+
+  key ||= undefined
+
+  if (key) {
+    if (key.byteLength < binding.crypto_generichash_KEYBYTES_MIN) throw new Error("key must be atleast 'crypto_generichash_KEYBYTES_MIN' bytes")
+    if (key.byteLength > binding.crypto_generichash_KEYBYTES_MAX) throw new Error("key must be at most 'crypto_generichash_KEYBYTES_MAX' bytes")
+  }
+
+  return module.exports._crypto_generichash_init(state, key, outputLength)
+}
+
+module.exports.crypto_generichash_batch = function (output, batch, key) {
+  if (isNode || batch.length < 12) {
+    // native-call uses argc=(2 | 3) to detect key presence
+    if (key) binding._crypto_generichash_batch(output, batch, key)
+    else binding._crypto_generichash_batch(output, batch)
```

Comment:
feels like you should just expose two functions then since this is private anyway? _crypto_generichash_batch_keyed and _crypto_generichash_batch

Reply:
if either of these related questions are time consuming, lets punt to 5.0.1 as a refactor for those

Reply:
The intention was to normalize the call signature to ensure that the call qualifies for fastpath regardless if key is present or not.

(It does not seem to matter if the 3rd arg is undefined or an buffer, but if argc reports 2; then the call automatically hits the slow path. Also, this way we trick the JIT to optimize a bit faster)

Reply:
lets fix in 5.0.1